sycl/source/detail/devicelib/glibc_wrapper.cpp

```diff
@@ -0,0 +1,26 @@
+//==--- glibc_wrapper.cpp - wrappers for Glibc internal functions ----------==//
+//
+// Part of the LLVM Project, under the Apache License v2.0 with LLVM Exceptions.
+// See https://llvm.org/LICENSE.txt for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+//
+//===----------------------------------------------------------------------===//
+
+#ifdef __SYCL_DEVICE_ONLY__
+#include "wrapper.h"
+
+#include <CL/__spirv/spirv_vars.hpp> // for __spirv_BuiltInGlobalInvocationId,
+                                     //     __spirv_BuiltInLocalInvocationId
+
+extern "C" SYCL_EXTERNAL
```

Comment:
Is it OK to add attribute SYCL_EXTERNAL at definition of function whilst there is no such attribute at declaration (in system header)?

Reply:
System header is not included here, so there is no declaration without SYCL_EXTERNAL.
The attribute is only needed to include the function into device code (since there is no kernel in this TU).

Reply:
Still the user #include-s system header that does where there is no SYCL_EXTERNAL in declaration of __assert_fail.

Reply:
SYCL 1.2.1 spec says:
> The default behavior in SYCL applications is that all the definitions and declarations of the functions and methods are available to the SYCL compiler, in the same translation unit. When this is not the case, all the symbols that need to be exported to a SYCL library or from a C++ library to a SYCL application need to be defined using the macro: SYCL_EXTERNAL. 

So the answer is "no" - this implementation does not strictly follow the SYCL spec.
There is no way we can patch system headers and declare standard functions with SYCL_EXTERNAL.

Reply:
I think it is unclear whether this sentence in the spec applies to anything other than user code. And because this patch effects a declaration in non-user code, it isn't a clear violation. What is important is that any such user or unsupported system declaration gives a FE error. We can do that by having a white-list for those supported system functions. (Not required in this patch - just a note for future work)